### ISSUE TITLE
Implemented addPolylines and removeAnnotations in JNI

### ISSUE THREAD

**Issue by ljbade** [mapbox/mapbox-gl-native#2809]
Currently `addPolylines` and `removeAnnotations` are implemented by looping through a list in Java.

We should improve these using JNI similar to `addMarkers` and `addPolygons`.

`removeAnnotations` in JNI will also help with removing the 10000 markers in the bulk marker test activity.

/cc @tobrun

**Comment by tobrun** [comment]
I can pick this up, good to learn some JNI while I'm at it.

**Comment by tobrun** [comment]
Current Java implementation of removeAnnotations:
```java
   /**
     * Removes multiple annotations from the map.
     *
     * @param annotationList A list of annotation objects to remove.
     */
    @UiThread
    public void removeAnnotations(@NonNull List<Annotation> annotationList) {
        if (annotationList == null) {
            throw new NullPointerException("annotationList is null");
        }

        // TODO make faster in JNI
        int count = annotationList.size();
        for (int i = 0; i < count; i++) {
            Annotation annotation = annotationList.get(i);
            removeAnnotation(annotation);
        }
    }
```

**Comment by tobrun** [comment]
There is already a bulk removeAnnotations method in NativeMapView, method will be refactored to:

```java
    /**
     * Removes multiple annotations from the map.
     *
     * @param annotationList A list of annotation objects to remove.
     */
    @UiThread
    public void removeAnnotations(@NonNull List<Annotation> annotationList) {
        if (annotationList == null) {
            throw new NullPointerException("annotationList is null");
        }
        
        long[]ids = new long[annotationList.size()];
        for (int i = 0; i < annotationList.size(); i++) {
            ids[i] = annotationList.get(i).getId();
        }
        mNativeMapView.removeAnnotations(ids);
    }
```

**Comment by tobrun** [comment]
Now with preallocation of size:

```
   /**
     * Removes multiple annotations from the map.
     *
     * @param annotationList A list of annotation objects to remove.
     */
    @UiThread
    public void removeAnnotations(@NonNull List<Annotation> annotationList) {
        if (annotationList == null) {
            throw new NullPointerException("annotationList is null");
        }

        int count = annotationList.size();
        long[]ids = new long[count];
        for (int i = 0; i < count; i++) {
            ids[i] = annotationList.get(i).getId();
        }
        mNativeMapView.removeAnnotations(ids);
    }
```

**Comment by tobrun** [comment]
Now taking a look at Polylines:
`addPolylines` doesn't have an equivalent in the form of `NativeMapView.addPolylines`. 
Will need to add one in `jni.cpp`

**Comment by tobrun** [comment]
JNI.cpp:

```c++
     {"nativeAddPolylines", "(JLjava/util/List;)[J",
      reinterpret_cast<void *>(&nativeAddPolylines)},
```

**Comment by tobrun** [comment]
Adjusted java impl:

```java
  /**
     * Adds multiple polylines to this map.
     *
     * @param polylineOptionsList A list of polyline options objects that defines how to render the polylines.
     * @return A list of the {@code Polyline}s that were added to the map.
     */
    @UiThread
    @NonNull
    public List<Polyline> addPolylines(@NonNull List<PolylineOptions> polylineOptionsList) {
        if (polylineOptionsList == null) {
            throw new NullPointerException("polylineOptionsList is null");
        }

        int count = polylineOptionsList.size();
        List<Polyline> polylines = new ArrayList<>(count);
        for (PolylineOptions options : polylineOptionsList) {
            polylines.add(options.getPolyline());
        }

        long[] ids = mNativeMapView.addPolylines(polylines);

        Polyline p;
        for (int i = 0; i < count; i++) {
            p = polylines.get(i);
            p.setId(ids[i]);
            p.setMapView(this);
            mAnnotations.add(p);
        }

        return new ArrayList<>(polylines);
    }
```

**Comment by tobrun** [comment]
Currently is crashes and burns with:

> 10-27 16:27:59.352 1926-1926/? W/dalvikvm: JNI WARNING: instance fieldID 0x45298e98 not valid for class Lcom/mapbox/mapboxsdk/annotations/Polyline;
10-27 16:27:59.352 1926-1926/? W/dalvikvm:              in Lcom/mapbox/mapboxsdk/views/NativeMapView;.nativeAddPolylines:(JLjava/util/List;)[J (GetIntField)

**Comment by tobrun** [comment]
Multiple builds later I have a working `nativeAddPolylines` method:

```c++
jlongArray JNICALL nativeAddPolylines(JNIEnv *env, jobject obj, jlong nativeMapViewPtr, jobject jlist) {
    mbgl::Log::Debug(mbgl::Event::JNI, "nativeAddPolylines");
    assert(nativeMapViewPtr != 0);
    NativeMapView *nativeMapView = reinterpret_cast<NativeMapView *>(nativeMapViewPtr);

    std::vector<mbgl::ShapeAnnotation> shapes;

    if (jlist == nullptr) {
        if (env->ThrowNew(nullPointerExceptionClass, "List cannot be null.") < 0) {
            env->ExceptionDescribe();
            return nullptr;
        }
        return nullptr;
    }

    jobjectArray jarray =
        reinterpret_cast<jobjectArray>(env->CallObjectMethod(jlist, listToArrayId));
    if (env->ExceptionCheck() || (jarray == nullptr)) {
        env->ExceptionDescribe();
        return nullptr;
    }

    jsize len = env->GetArrayLength(jarray);
    if (len < 0) {
        env->ExceptionDescribe();
        return nullptr;
    }

    shapes.reserve(len);

    for (jsize i = 0; i < len; i++) {
        jobject polyline = reinterpret_cast<jobject>(env->GetObjectArrayElement(jarray, i));

        jfloat alpha = env->GetFloatField(polyline, polylineAlphaId);
        if (env->ExceptionCheck()) {
            env->ExceptionDescribe();
            return nullptr;
        }

        jint color = env->GetIntField(polyline, polylineColorId);
        if (env->ExceptionCheck()) {
            env->ExceptionDescribe();
            return nullptr;
        }

        int r = (color >> 16) & 0xFF;
        int g = (color >> 8) & 0xFF;
        int b = (color) & 0xFF;
        int a = (color >> 24) & 0xFF;

        jfloat width = env->GetFloatField(polyline, polylineWidthId);
        if (env->ExceptionCheck()) {
            env->ExceptionDescribe();
            return nullptr;
        }

        mbgl::ShapeAnnotation::Properties shapeProperties;
        mbgl::LinePaintProperties lineProperties;
        lineProperties.opacity = alpha;
        lineProperties.color = {{ static_cast<float>(r) / 255.0f, static_cast<float>(g) / 255.0f, static_cast<float>(b) / 255.0f, static_cast<float>(a) / 255.0f }};
        lineProperties.width = width;
        shapeProperties.set<mbgl::LinePaintProperties>(lineProperties);

        jobject points = env->GetObjectField(polyline, polylinePointsId);
        mbgl::AnnotationSegment segment = annotation_segment_from_latlng_jlist(env, points);

        shapes.emplace_back(mbgl::AnnotationSegments { segment }, shapeProperties);

        env->DeleteLocalRef(polyline);
    }

    env->DeleteLocalRef(jarray);

    std::vector<uint32_t> shapeAnnotationIDs = nativeMapView->getMap().addShapeAnnotations(shapes);
    return std_vector_uint_to_jobject(env, shapeAnnotationIDs);
}
```

**Comment by tobrun** [comment]
Closed with  f3ffb45

**Comment by ljbade** [comment]
@tobrun Nice work! It's always good to have everyone here at least try some JNI to understand the pain and complexity it adds over say iOS.
